consul/prepared_query/template.go

```diff
@@ -0,0 +1,186 @@
+package prepared_query
+
+import (
+	"fmt"
+	"reflect"
+	"regexp"
+	"strings"
+
+	"github.com/hashicorp/consul/consul/structs"
+	"github.com/hashicorp/hil"
+	"github.com/hashicorp/hil/ast"
+	"github.com/mitchellh/copystructure"
+)
+
+// IsTemplate returns true if the given query is a template.
+func IsTemplate(query *structs.PreparedQuery) bool {
+	return query.Template.Type != ""
+}
+
+// CompiledTemplate is an opaque object that can be used later to render a
+// prepared query template.
+type CompiledTemplate struct {
+	// query keeps a copy of the original query for rendering.
+	query *structs.PreparedQuery
+
+	// trees contains a map with paths to string fields in a structure to
+	// parsed syntax trees, suitable for later evaluation.
+	trees map[string]ast.Node
+
+	// re is the compiled regexp, if they supplied one (this can be nil).
+	re *regexp.Regexp
+}
+
+// Compile validates a prepared query template and returns an opaque compiled
+// object that can be used later to render the template.
+func Compile(query *structs.PreparedQuery) (*CompiledTemplate, error) {
+	// Make sure it's a type we understand.
+	if query.Template.Type != structs.QueryTemplateTypeNamePrefixMatch {
+		return nil, fmt.Errorf("Bad Template.Type '%s'", query.Template.Type)
+	}
+
+	// Start compile.
+	ct := &CompiledTemplate{
+		trees: make(map[string]ast.Node),
+	}
+
+	// Make a copy of the query to use as the basis for rendering later.
+	dup, err := copystructure.Copy(query)
+	if err != nil {
+		return nil, err
+	}
+	var ok bool
+	ct.query, ok = dup.(*structs.PreparedQuery)
+	if !ok {
+		return nil, fmt.Errorf("Failed to copy query")
+	}
+
+	// Walk over all the string fields in the Service sub-structure and
+	// parse them as HIL.
+	parse := func(path string, v reflect.Value) error {
+		tree, err := hil.Parse(v.String())
+		if err != nil {
+			return fmt.Errorf("Bad format '%s' in Service%s: %s", v.String(), path, err)
```

Comment:
Yeah - this is a little cheesy but all paths have a leading "." so this looks like `Service.Foo`.